src/envoy/http/service_control/filter_fuzz_test.cc

```diff
@@ -0,0 +1,212 @@
+      throw ProtoValidationException("num sidestream", input);
+    }
+
+    // Setup mocks.
+    NiceMock<MockFactoryContext> context;
+    NiceMock<Http::MockAsyncClientRequest> response(
+        &context.cluster_manager_.async_client_);
+    NiceMock<Envoy::Http::MockStreamDecoderFilterCallbacks>
+        mock_decoder_callbacks;
+    NiceMock<Envoy::Http::MockStreamEncoderFilterCallbacks>
+        mock_encoder_callbacks;
+
+    // Return a fake span.
+    EXPECT_CALL(mock_decoder_callbacks, activeSpan())
+        .WillRepeatedly(ReturnRef(Envoy::Tracing::NullSpan::instance()));
+
+    // Callback for token subscriber to start.
+    Envoy::Event::TimerCb onReadyCallback;
+    EXPECT_CALL(context.dispatcher_, createTimer_(_))
+        .WillRepeatedly(
+            Invoke([&onReadyCallback](const Envoy::Event::TimerCb& cb) {
+              ENVOY_LOG_MISC(trace, "Mocking dispatcher createTimer");
+              onReadyCallback = cb;
+              return new NiceMock<Event::MockTimer>();
+            }));
+
+    // Mock the http async client.
+    int resp_num = 0;
+    EXPECT_CALL(context.cluster_manager_.async_client_, send_(_, _, _))
+        .WillRepeatedly(Invoke([&response, &input, &resp_num](
+                                   const Envoy::Http::RequestMessagePtr&,
+                                   Envoy::Http::AsyncClient::Callbacks&
+                                       callback,
+                                   const Envoy::Http::AsyncClient::
+                                       RequestOptions&) {
+          // FIXME(nareddyt): For now, just increment the counter for response
+          // numbers.
+          auto& response_data = input.sidestream_response().Get(
+              resp_num++ % input.sidestream_response().size());
+
+          // Create the response message.
+          auto headers =
+              Envoy::Fuzz::fromHeaders<Envoy::Http::TestResponseHeaderMapImpl>(
+                  response_data.headers());
+          auto headers_ptr =
+              std::make_unique<Envoy::Http::TestResponseHeaderMapImpl>(headers);
+          auto trailers =
+              Envoy::Fuzz::fromHeaders<Envoy::Http::TestResponseTrailerMapImpl>(
+                  response_data.trailers());
+          auto trailers_ptr =
+              std::make_unique<Envoy::Http::TestResponseTrailerMapImpl>(
+                  trailers);
+
+          auto msg = std::make_unique<Envoy::Http::ResponseMessageImpl>(
+              std::move(headers_ptr));
+          msg->trailers(std::move(trailers_ptr));
+          msg->body() =
+              std::make_unique<Buffer::OwnedImpl>(response_data.data().Get(0));
+
+          // Callback.
+          callback.onSuccess(std::move(msg));
+          return &response;
+        }));
+
+    // Fuzz the stream info.
+    TestStreamInfo stream_info =
+        Envoy::Fuzz::fromStreamInfo(input.stream_info());
+    EXPECT_CALL(mock_decoder_callbacks, streamInfo())
+        .WillRepeatedly(ReturnRef(stream_info));
```

Comment:
Instead of setting expectation (slow-down), can you set `mock_decoder_callbacks.stream_info_ = stream_info`, the expectation to return `stream_info_` is already made in initialization. Does this speed-up?

Reply:
Unfortunately, this didn't work due to mocks vs fakes.

```
src/envoy/http/service_control/filter_fuzz_test.cc:182:41: error: no match for 'operator=' (operand types are 'testing::NiceMock<Envoy::StreamInfo::MockStreamInfo>' and 'Envoy::TestStreamInfo')
  182 |   mock_decoder_callbacks.stream_info_ = stream_info;
```

Reply:
Ahhhh. Gotcha. Didn't realize `stream_info_` was a mock. Boo.

I think this is fine as is as a start, and you can iterat for speed ups. I think it'll be possible to make some of the mocks and expectations static and updating/cleaning up per-run, but it's a good task to iterate on.

Reply:
Good point, could you share an example of this in Envoy? I'm not familiar with how this will speed up execution.

Yup, I can iterate on this next week. I'd like to get this merged first so OSS-Fuzz will pick it up and display some initial fuzz stats.

Reply:
Sounds like a plan!

Reply:
As discussed offline: https://github.com/envoyproxy/envoy/pull/8397